File: class_primer/class_primer__super_super_supercar.cs
using System.Linq;
using static System.Console;

internal class Program
{
    abstract class Car
    {
        internal int Cons { get; set; }
        internal int Fuel { get; set; }
        internal int Len { get; set; } = 0;

        protected Car( int f, int c )
        {
            Cons = c;
            Fuel = f;
        }

        internal abstract void Fly();
        internal abstract void Run();
        internal abstract void Telepo();
    }

    class SC : Car
    {
        internal SC( int f, int c ) : base( f, c ) { }

        internal override void Run()
        {
            if( 0 < Fuel ) {
                Fuel--;
                Len += Cons;
            }
            return;
        }

        internal override void Fly() { }
        internal override void Telepo() { }
    }

    class SSC : SC
    {
        internal SSC( int f, int c ) : base( f, c ) { }

        internal override void Fly()
        {
            if( 5 <= Fuel ) {
                Fuel -= 5;
                Len += Cons * Cons;
            } else {
                Run();
            }
            return;
        }
    }

    class UC : SSC
    {
        internal UC( int f, int c ) : base( f, c ) { }

        internal override void Fly()
        {
            if( 5 <= Fuel ) {
                Fuel -= 5;
                Len += 2 * Cons * Cons;
            } else {
                Run();
            }
            return;
        }

        internal override void Telepo()
        {
            if( Cons * Cons <= Fuel ) {
                Fuel -= Cons * Cons;
                Len += (int)System.Math.Pow( Cons, 4d );
            } else {
                Fly();
            }
            return;
        }
    }

    static void Main()
    {
        int[] n = ReadLine().Split().Select( int.Parse ).ToArray();
        Car[] a = new Car[n[0]];

        foreach( int i in Enumerable.Range( 0, n[0] ) ) {
            string[] t = ReadLine().Split();

            switch( t[0] ) {
            case "supercar":
                a[i] = new SC( int.Parse( t[1] ), int.Parse( t[2] ) );
                break;
            case "supersupercar":
                a[i] = new SSC( int.Parse( t[1] ), int.Parse( t[2] ) );
                break;
            case "supersupersupercar":
                a[i] = new UC( int.Parse( t[1] ), int.Parse( t[2] ) );
                break;
            }
        }
        foreach( int _ in Enumerable.Range( 0, n[1] ) ) {
            string[] t = ReadLine().Split();

            switch( t[1] ) {
            case "run":
                a[int.Parse( t[0] ) - 1].Run();
                break;
            case "fly":
                a[int.Parse( t[0] ) - 1].Fly();
                break;
            case "teleport":
                a[int.Parse( t[0] ) - 1].Telepo();
                break;
            }
        }
        SetOut( new System.IO.StreamWriter( OpenStandardOutput() ) { AutoFlush = false } );
        foreach( Car c in a ) {
            WriteLine( c.Len );
        }
        Out.Flush();
        return;
    }
}

File: class_primer/class_primer__super_super_supercar.py
class car:
    def __init__(self, f, c):
        self.cons = c
        self.fuel = f
        self.len = 0

    def getLen(self):
        return self.len

class sc(car):
    def __init__(self, f, c):
        super().__init__(f, c)

    def run(self):
        if 0 < self.fuel:
            self.fuel -= 1
            self.len += self.cons

class ssc(sc):
    def __init__(self, f, c):
        super().__init__(f, c)

    def fly(self):
        if 5 <= self.fuel:
            self.fuel -= 5
            self.len += self.cons * self.cons
        else:
            self.run()

class uc(ssc):
    def __init__(self, f, c):
        super().__init__(f, c)

    def fly(self):
        if 5 <= self.fuel:
            self.fuel -= 5
            self.len += 2 * self.cons * self.cons
        else:
            self.run()

    def telepo(self):
        if self.cons * self.cons <= self.fuel:
            self.fuel -= self.cons * self.cons
            self.len += pow(self.cons, 4)
        else:
            self.fly()

n, k = list(map(int, input().split()))
l = []

for _ in range(n):
    t = input().split()

    if t[0] == "supercar":
        l.append(sc(int(t[1]), int(t[2])))
    elif t[0] == "supersupercar":
        l.append(ssc(int(t[1]), int(t[2])))
    elif t[0] == "supersupersupercar":
        l.append(uc(int(t[1]), int(t[2])))
for _ in range(k):
    t = input().split()

    if t[1] == "run":
        l[int(t[0]) - 1].run()
    elif t[1] == "fly":
        l[int(t[0]) - 1].fly()
    elif t[1] == "teleport":
        l[int(t[0]) - 1].telepo()
for i in l:
    print(i.getLen())
